File: leetcode/cpp/39CombinationSum.cpp
#include <vector>
using namespace std;

class Solution {
    vector<vector<int>> result;
    vector<int> current;
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        /*
            so can we have for exampekl 2 chsoen as many tiems as we want as long as it == target?
            so  if we target is 7 and we have combination 1,1,1,1,1,1,1; this would be not allowed because the frequency of the nums is different.
        */

                    
        backtrack(candidates, 0, target);
        return result;
    }
    void backtrack(vector<int>& candidates, int start, int remaining){

        if(remaining == 0){
            result.push_back(current);
            return;
        }

        if(remaining < 0) return;

        for(int i = start; i < candidates.size(); i++){
            current.push_back(candidates[i]);
            backtrack(candidates, i, remaining - candidates[i]);
            current.pop_back();
        }
    }
};

File: leetcode/python/39CombinationSum.py
# 39. Combination Sum
# Link: https://leetcode.com/problems/combination-sum/
from typing import List

class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        
        res = []
        self.backtrack(nums, 0, [], target, res)
        return res

    def backtrack(self, nums, start, subset, remaining, res):
        if remaining == 0:
            res.append(subset[:])
            return

        for i in range(start, len(nums)):
            if nums[i] > remaining:
                continue

            subset.append(nums[i])
            self.backtrack(nums, i, subset, remaining - nums[i], res)
            subset.pop()

if __name__ == "__main__":
    s = Solution()
    print(s.combinationSum([2, 3, 6, 7], 7))